Assertions on nested comprehensions can give NameError

This test:

```
def test_nested_comprehension():
    s = [[True, True], [True, True]]
    assert all(e for row in s for e in row)
```
works fine in pytest 4.4.1 and 4.5.0.  In pytest 4.6.1 it produces a NameError:

```
$ pytest
===================================================== test session starts =====================================================
platform darwin -- Python 3.7.3, pytest-4.6.1, py-1.8.0, pluggy-0.12.0
collected 1 item                                                                                                              

test_it.py F                                                                                                            [100%]

========================================================== FAILURES ===========================================================
__________________________________________________ test_nested_comprehension __________________________________________________

    def test_nested_comprehension():
        s = [[True, True], [True, True]]
>       assert all(e for row in s for e in row)
E       NameError: name 'e' is not defined

test_it.py:3: NameError
================================================== 1 failed in 0.06 seconds ===================================================
```

Pip list gives:
```
$ pip list
Package            Version
------------------ -------
atomicwrites       1.3.0  
attrs              19.1.0 
importlib-metadata 0.17   
more-itertools     7.0.0  
packaging          19.0   
pip                19.1.1 
pluggy             0.12.0 
py                 1.8.0  
pyparsing          2.4.0  
pytest             4.6.1  
setuptools         41.0.1 
six                1.12.0 
wcwidth            0.1.7  
wheel              0.33.4 
zipp               0.5.1  
```

--- a/changelog/5358.bugfix.rst
+++ b/changelog/5358.bugfix.rst
@@ -1,1 +1,0 @@
-Fix assertion rewriting of ``all()`` calls to deal with non-generators.

--- a/changelog/5370.bugfix.rst
+++ b/changelog/5370.bugfix.rst
@@ -1,0 +1,1 @@
+Revert unrolling of ``all()`` to fix ``NameError`` on nested comprehensions.

--- a/changelog/5371.bugfix.rst
+++ b/changelog/5371.bugfix.rst
@@ -1,0 +1,1 @@
+Revert unrolling of ``all()`` to fix incorrect handling of generators with ``if``.

--- a/changelog/5372.bugfix.rst
+++ b/changelog/5372.bugfix.rst
@@ -1,0 +1,1 @@
+Revert unrolling of ``all()`` to fix incorrect assertion when using ``all()`` in an expression.

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -903,21 +903,11 @@ warn_explicit(
         res = self.assign(ast.BinOp(left_expr, binop.op, right_expr))
         return res, explanation
 
-    @staticmethod
-    def _is_any_call_with_generator_or_list_comprehension(call):
-        """Return True if the Call node is an 'any' call with a generator or list comprehension"""
-        return (
-            isinstance(call.func, ast.Name)
-            and call.func.id == "all"
-            and len(call.args) == 1
-            and isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp))
-        )
-
     def visit_Call(self, call):
         """
         visit `ast.Call` nodes
         """
-        if self._is_any_call_with_generator_or_list_comprehension(call):
+        if isinstance(call.func, ast.Name) and call.func.id == "all":
             return self._visit_all(call)
         new_func, func_expl = self.visit(call.func)
         arg_expls = []
@@ -944,6 +934,8 @@ warn_explicit(
 
     def _visit_all(self, call):
         """Special rewrite for the builtin all function, see #5062"""
+        if not isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp)):
+            return
         gen_exp = call.args[0]
         assertion_module = ast.Module(
             body=[ast.Assert(test=gen_exp.elt, lineno=1, msg="", col_offset=1)]

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -907,8 +907,6 @@ warn_explicit(
         """
         visit `ast.Call` nodes
         """
-        if isinstance(call.func, ast.Name) and call.func.id == "all":
-            return self._visit_all(call)
         new_func, func_expl = self.visit(call.func)
         arg_expls = []
         new_args = []
@@ -931,27 +929,6 @@ warn_explicit(
         res_expl = self.explanation_param(self.display(res))
         outer_expl = "{}\n{{{} = {}\n}}".format(res_expl, res_expl, expl)
         return res, outer_expl
-
-    def _visit_all(self, call):
-        """Special rewrite for the builtin all function, see #5062"""
-        if not isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp)):
-            return
-        gen_exp = call.args[0]
-        assertion_module = ast.Module(
-            body=[ast.Assert(test=gen_exp.elt, lineno=1, msg="", col_offset=1)]
-        )
-        AssertionRewriter(module_path=None, config=None).run(assertion_module)
-        for_loop = ast.For(
-            iter=gen_exp.generators[0].iter,
-            target=gen_exp.generators[0].target,
-            body=assertion_module.body,
-            orelse=[],
-        )
-        self.statements.append(for_loop)
-        return (
-            ast.Num(n=1),
-            "",
-        )  # Return an empty expression, all the asserts are in the for_loop
 
     def visit_Starred(self, starred):
         # From Python 3.5, a Starred node can appear in a function call

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -635,12 +635,6 @@ class TestAssertionRewrite:
         else:
             assert lines == ["assert 0 == 1\n +  where 1 = \\n{ \\n~ \\n}.a"]
 
-    def test_unroll_expression(self):
-        def f():
-            assert all(x == 1 for x in range(10))
-
-        assert "0 == 1" in getmsg(f)
-
     def test_custom_repr_non_ascii(self):
         def f():
             class A:
@@ -655,53 +649,6 @@ class TestAssertionRewrite:
         msg = getmsg(f)
         assert "UnicodeDecodeError" not in msg
         assert "UnicodeEncodeError" not in msg
-
-    def test_unroll_generator(self, testdir):
-        testdir.makepyfile(
-            """
-            def check_even(num):
-                if num % 2 == 0:
-                    return True
-                return False
-
-            def test_generator():
-                odd_list = list(range(1,9,2))
-                assert all(check_even(num) for num in odd_list)"""
-        )
-        result = testdir.runpytest()
-        result.stdout.fnmatch_lines(["*assert False*", "*where False = check_even(1)*"])
-
-    def test_unroll_list_comprehension(self, testdir):
-        testdir.makepyfile(
-            """
-            def check_even(num):
-                if num % 2 == 0:
-                    return True
-                return False
-
-            def test_list_comprehension():
-                odd_list = list(range(1,9,2))
-                assert all([check_even(num) for num in odd_list])"""
-        )
-        result = testdir.runpytest()
-        result.stdout.fnmatch_lines(["*assert False*", "*where False = check_even(1)*"])
-
-    def test_for_loop(self, testdir):
-        testdir.makepyfile(
-            """
-            def check_even(num):
-                if num % 2 == 0:
-                    return True
-                return False
-
-            def test_for_loop():
-                odd_list = list(range(1,9,2))
-                for num in odd_list:
-                    assert check_even(num)
-        """
-        )
-        result = testdir.runpytest()
-        result.stdout.fnmatch_lines(["*assert False*", "*where False = check_even(1)*"])
 
 
 class TestRewriteOnImport:

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -656,7 +656,7 @@ class TestAssertionRewrite:
         assert "UnicodeDecodeError" not in msg
         assert "UnicodeEncodeError" not in msg
 
-    def test_unroll_all_generator(self, testdir):
+    def test_unroll_generator(self, testdir):
         testdir.makepyfile(
             """
             def check_even(num):
@@ -671,7 +671,7 @@ class TestAssertionRewrite:
         result = testdir.runpytest()
         result.stdout.fnmatch_lines(["*assert False*", "*where False = check_even(1)*"])
 
-    def test_unroll_all_list_comprehension(self, testdir):
+    def test_unroll_list_comprehension(self, testdir):
         testdir.makepyfile(
             """
             def check_even(num):
@@ -685,31 +685,6 @@ class TestAssertionRewrite:
         )
         result = testdir.runpytest()
         result.stdout.fnmatch_lines(["*assert False*", "*where False = check_even(1)*"])
-
-    def test_unroll_all_object(self, testdir):
-        """all() for non generators/non list-comprehensions (#5358)"""
-        testdir.makepyfile(
-            """
-            def test():
-                assert all((1, 0))
-            """
-        )
-        result = testdir.runpytest()
-        result.stdout.fnmatch_lines(["*assert False*", "*where False = all((1, 0))*"])
-
-    def test_unroll_all_starred(self, testdir):
-        """all() for non generators/non list-comprehensions (#5358)"""
-        testdir.makepyfile(
-            """
-            def test():
-                x = ((1, 0),)
-                assert all(*x)
-            """
-        )
-        result = testdir.runpytest()
-        result.stdout.fnmatch_lines(
-            ["*assert False*", "*where False = all(*((1, 0),))*"]
-        )
 
     def test_for_loop(self, testdir):
         testdir.makepyfile(

